grpc.go

```diff
@@ -1449,3 +1449,26 @@ func (a *agentGRPC) SetGuestDateTime(ctx context.Context, req *pb.SetGuestDateTi
 	}
 	return &gpb.Empty{}, nil
 }
+
+func (a *agentGRPC) CopyFile(ctx context.Context, req *pb.CopyFileRequest) (*gpb.Empty, error) {
+	if err := os.MkdirAll(filepath.Dir(req.Path), os.FileMode(req.DirMode)); err != nil {
+		return emptyResp, err
+	}
+
+	file, err := os.OpenFile(req.Path, os.O_WRONLY|os.O_CREATE, os.FileMode(req.FileMode))
+	if err != nil {
+		return emptyResp, err
+	}
+
+	defer file.Close()
+
+	if _, err := file.Seek(req.Offset, 0); err != nil {
+		return emptyResp, err
+	}
+
+	if _, err := file.Write(req.Content); err != nil {
```

Comment:
It would be safer to write the copied file to a temporary one, then atomically move it over the top of the existing one, overwriting the original. The downside being we'll temporarily need more disk space to retain two copies. We'll also have to take care to ensure the temp copy is written in the same directory as the original to avoid any FS-crossing performance issues.

Reply:
sounds good, but then we'll need two commands `CopyFile` and `MoveFile` because `CopyFile` doesn't know when the temporarily file should be renamed/moved

Reply:
I just mean that the function could do something like the following (untested):

```go
dir := filepath.Dir(req.Path)
file := filepath.Base(req.Path)

// create hidden temp file in same directory as original file with a random suffix
tmpfile, err := ioutil.TempFile(dir, "."+file+"*")

name := tmpfile.Name()
tmpfile.WriteAt(...)
tmpfile.Close()

// move the temp file over the top of the destination path atomically
err := os.Rename(name, req.Path)
```

Reply:
@jodh-intel  ohh I see a problem with this design, if offset is `!= 0` and `dest` exist, it must be copied to the temporary file, the problem is when `dest` is a humongous file, copying `dest` to the temporarily file will be expensive in both time and disk space.

Reply:
yep - slow, but safe. We need to capture these sorts of scenarios in a doc and decide on the best approach in each case I think.

Reply:
@sboeuf writes:
well you need to think about this CopyFile() command as a workaround for 9p, so we're not planning on having huge files being passed through here. Thus, don't know if we should worry about performance here. I value more the security check personally.

Reply:
I have changed the algorithm to copy files, please take a look

Reply:
Is this comment properly addressed? Can we resolve it?